### DISCUSSION TITLE
how to exclude test file for diktat plugin, but enabling it for ktlint plugin.

### DISCUSSION THREAD

**Discussion by Raghunath-SK** [diffplug/spotless#2339] [Q&A]
Hi team,

I'm using the spotless plugin in my gradle kotlin project.  I have used the ktlint and diktat from the spotless. 
I have a situation where I need to avoid running the diktat for test files , but run only for src files. 
But for ktlint, it has to run on both src and test files. 

I have made the below configuration in my build.gradle.kts , but seems like both ktlint & diktat is not happening for test files. 
(I want only ktlint to run for the test files)

```
spotless {
	kotlin {
            ktlint().apply {
                target("src/main/kotlin/**/*.kt")
                editorConfigOverride(
                mapOf(
                    "ij_kotlin_allow_trailing_comma" to "true",
                    "ij_kotlin_allow_trailing_comma_on_call_site" to "true",
                    "ktlint_function_naming_ignore_when_annotated_with" to "Composable"
                )
            )
            }
            diktat().apply {
                configFile("./../diktat-analysis.yml")
                targetExclude("src/test/kotlin/**/*.kt")
            }
	}
		kotlinGradle {
			diktat().configFile("./../diktat-analysis.yml")
		}
}
```

WE can clearly see that I have a  targetExclude just for diktat(). 

Can anyone help me here please ? Am I missing something ?

**Comment by nedtwigg** [comment]
On GitHub, use triple-backticks for multi-line code blocks (I edited your question above).

The `kotlin` block defines ONE set of files and ONE list of steps. You can't change the files per-step.

If you want, you can do this:

```
spotles {
  kotlin {
    target("src/main/kotlin/**/*.kt")
    targetExclude('src/test/kotlin/**/*.kt')
    ktlint()
    diktat()
  }
  format 'kotlin2', com.diffplug.gradle.spotless.KotlinExtension {
    target("src/test/kotlin/**/*.kt")
    ktlint()
  }
}
```

but it's kinda icky. I don't recommend having any files get formatted by two different formats.

**Reply by Raghunath-SK** [reply]
Hi @nedtwigg , 

Thanks for the response.
I have Tried the similar code already. 
Also tried the same code from the above answer. Still its not working as expected. Its ignoring to run the ktlint and diktat for the src files now. Looks like the latest target is getting picked everytime. 


Coming to the main reason why I wanted to avoid diktat for test files.  We have a monorepo style multiple module project. But we will run spotless for individual module when doing changes to the particular module.  For some reason spotlessApply (when diktat is enabled) it is taking ages to complete the spotlessApply. We noticed, it was running very fast for src files, but it ran very slow for test files (sometimes 20+mins). 

This was the reason we are trying to ignore the diktat from the spotless, just for test files.
